test_app/tests/authentication/test_backend.py

```diff
@@ -172,3 +172,163 @@ def authenticate(*args, **kwars):
         # Expect the log we emit
         with expected_log('ansible_base.authentication.backend.logger', "exception", "Exception raised while trying to authenticate with"):
             backend.AnsibleBaseAuth().authenticate(None)
+
+
+@pytest.mark.django_db
+def test_last_login_from_with_attribute(local_authenticator, random_user, expected_log):
+    """Test that last_login_from is set when user has the attribute."""
+    # Add last_login_from attribute to user
+    random_user.last_login_from = None
+
+    # Create a mock authenticator plugin object with proper structure
+    mock_authenticator_plugin = mock.MagicMock()
+    mock_authenticator_plugin.authenticate.return_value = random_user
+    mock_authenticator_plugin.database_instance = local_authenticator
+
+    # Mock the save method to track calls
+    with mock.patch.object(random_user, 'save') as mock_save:
+        with mock.patch(
+            "ansible_base.authentication.backend.get_authentication_backends",
+            return_value={local_authenticator.id: mock_authenticator_plugin},
+        ):
+            # Expected log message when user logs in
+            with expected_log(
+                'ansible_base.authentication.backend.logger',
+                "info",
+                f'User {random_user.username} logged in from authenticator with ID "{local_authenticator.id}"',
+            ):
+                auth_return = backend.AnsibleBaseAuth().authenticate(None)
+
+            # Verify the user is returned
+            assert auth_return == random_user
+
+            # Verify last_login_from was set to the authenticator database instance
+            assert random_user.last_login_from == local_authenticator
+
+            # Verify save was called with the correct update_fields
+            mock_save.assert_called_once_with(update_fields=['last_login_from'])
+
+
+@pytest.mark.django_db
+def test_last_login_from_without_attribute(local_authenticator, random_user):
+    """Test that authentication works normally when user doesn't have last_login_from attribute."""
+    # Ensure user doesn't have last_login_from attribute
```

Comment:
Yeah in this case I think we should get the user logged in message as well since we always did before.